### ISSUE TITLE
Step3 is not working.

### ISSUE THREAD

**Issue by jense-arntz** [vivek1794/webrtc-android-codelab#10]
I have run the step3 application on Chrome Browser and Android app. Localvideo is working and remote video is not working.... what I missed?

**Comment by vivek1794** [comment]
The step-3 browser code may not be right. Check how the data is transferred (SDP and ice candidates). I did not concentrate on the web part and copied it from the webrtc codelabs. Android app code might be sending the data in a different format (JSON) than what the web is expecting. You can make the required changes to the js file to parse the data. Send me a PR if you made it work.

**Comment by jense-arntz** [comment]
Thanks Vivek. as you mentioned above, android code will be working properly? I only need to make the browser code working?

**Comment by jense-arntz** [comment]
Hi Vivek

**Comment by vivek1794** [comment]
Hi @jense-arntz yes the android to android part would be working. You need to make some changes to the browser code to make it work for Android <-> browser

**Comment by jense-arntz** [comment]
- Issue on STEP2.
when I build the step2 for testing android <-> android, there  is error on CustomCameraEventsHandler.java file and CustomPeerConnectionObserver.java file.

Error is **"Class CustomCameraEventsHandler.java must either be declared abstract or implement abstract method 'onCameraOpening(String)' in 'CameraEventsHandler'**

-Issue on STEP3.
There is no error occurred on Browser and Android when building peerconnection and completing the process 'Offer' and 'Answer'.
I am suspicious if there is issue on iceserver.

Please help. thanks.

**Comment by jense-arntz** [comment]
Please give me some hints.

**Comment by vivek1794** [comment]
@jense-arntz : The step-2 needs to be updated. Please refer to step-3 code for how to handle peer connection. WebRTC has removed many of the methods used in Step-2. 

Step-3 browser to android would not work as I said earlier. Browser side code needs to be changed to accept offer and answer signalling messages in JSON format just as the Android client would expect.

**Comment by darwish1994** [comment]
as i understand from tutorial ( part 1 -- 4 ) the single server in that project is sokect io server 
i make a local server like this https://codelabs.developers.google.com/codelabs/webrtc-web/#6 
and replace the url (" your_socket_io_instance_url_with_port") in SignallingClient class
 but remote vedio not work

**Comment by paveltech** [comment]
![image](https://user-images.githubusercontent.com/7562445/41767679-2463be00-762c-11e8-9bb2-9701fd69fc50.png)


This constructor gives me error. How you solved this error??

**Comment by arfa-maddi** [comment]
change build.gradle to this : implementation 'org.webrtc:google-webrtc:1.0.22672'
to solve constructore and VideoRenderer error

**Comment by skumarm** [comment]
> 1.0.22672

Yes. this fixed my issue on **VideoRenderer error**

**Comment by ArjunHalli** [comment]
Appliaction is crashing on launch and gives the error 
xyz.vivekc.webrtccodelab E/AndroidRuntime: FATAL EXCEPTION: main
    Process: xyz.vivekc.webrtccodelab, PID: 11587
    java.lang.NullPointerException
        at xyz.vivekc.webrtccodelab.MainActivity$1.onResponse(MainActivity.java:106)

**Comment by neilyoung** [comment]
@arfa-maddi You could create your own key.pem and cert using openssl for creating self signed certificate.
@skumarm I think this is because the request to obtain the TURN server fails in JS

```
https://service.xirsys.com/ice?ident=vivekchanddru&secret=ad6ce53a-e6b5-11e6-9685-937ad99985b9&domain=www.vivekc.xyz&application=default&room=testing&secure=1
```
If your signaller runs at localhost, the iceserver configuration is NULL. The Java code seems to be not prepared for this and crashes.

Anyway, the a.m. URL seems to not work any longer. You could try to obtain a free TURN instance at NUMB http://numb.viagenie.ca/ and hard code the URL and credentials in signalling/js/main.js like so:

  // Hardcoded
    pcConfig.iceServers.push({
      'urls': 'stun:<ip>:3478'
    })

    pcConfig.iceServers.push({
      'urls': 'turn:<ip>:3478',
      'username': 'username',
      'credential': 'password'
    })

**Comment by vivek1794** [comment]
I did not find the time to update the project. Please raise a PR so that we can keep it updated. 

Also, WebRTC changes their class names and methods so often that it is difficult to keep it updated for the current version. It should, however, work with the AAR file attached in this repo.

**Comment by chutongz** [comment]
> @arfa-maddi You could create your own key.pem and cert using openssl for creating self signed certificate.


@neilyoung Hi, I am having issue about key.pem and cer.pem. Should I just write the id_rsa and id_rsa.pub location (which ssh generated when the git configuration) down to replace key.pem and cer.pem?

**Comment by accuware** [comment]
@chutongz Your signaling server needs to be available as secure server. For that you need to either provide a "real" certificate or you can create your own self-signed certificate. There are tons of articles on the web regarding "How to create self-signed certificate using openssl". However, modern browsers meanwhile are making problems accepting a self-signed certificate, so in case you need to convince browsers that you are a trusted CA you need to do a bit more (or to take care about a little bit more) the time you are creating your self-singed certificate. This is excellently explained in this post https://fabianlee.org/2018/02/17/ubuntu-creating-a-trusted-ca-and-san-certificate-using-openssl-on-ubuntu/

**Comment by chutongz** [comment]
@accuware Hi, thank you so much for replying me!
Now I used openssl to create the key.pem and cert.pem already, but when I try the command "node index.js", it gave me error as below:
![tim 20181203160725](https://user-images.githubusercontent.com/30098450/49360934-94587880-f715-11e8-8625-c0e98041eff3.png)
 Do you have any idea of this error? Is it because some thing wrong when I create the certification?

**Comment by neilyoung** [comment]
@chutongz Sorry, used the wrong account. It was me :)
There is for sure a format problem. The cert is expected to be Base64 text, IMHO

**Comment by chutongz** [comment]
@neilyoung Yes, I recreate the certification and run the command but nothing happend...
Nothing returned, nothing printed... I waited for so long....
I am so confused and wondering if it comes up with the socket.io url? Because I haven't set it yet...
![tim 20181203164934](https://user-images.githubusercontent.com/30098450/49363006-74c44e80-f71b-11e8-95ab-b9612a44a391.png)

**Comment by neilyoung** [comment]
Well, I'm neither the creator nor a contributor of this app....

Once you have setup the server you should be able to test the connection using

`telnet your.server 443``

or whatever port the server is listening to.

**Comment by chutongz** [comment]
@neilyoung OK! I'll try it anyway. Thanks so much for replying me again! :)

**Comment by mmarashan** [comment]
Hello everyone! Check that
`remoteVideoTrack.addRenderer(new VideoRenderer(remoteVideoSurfaceViewRender));`
or
`remoteVideoTrack.addSink(remoteVideoSurfaceViewRender);`
call after ALL ICE exchange occurred.
It depends on signalling and can cause a heavily detectable error
